src/analysis/modal-utils.ts

```diff
@@ -33,61 +39,136 @@ export interface IComputeConfig {
+  workerDiskSize?: number | null | undefined;
+  timeoutInMinutes?: number | null;
+};
+
+// Auxiliary functions -- begin
+export const isGce = (runtimeType) => runtimeType === runtimeTypes.gceVm;
+
+export const isDataproc = (runtimeType) =>
+  runtimeType === runtimeTypes.dataprocSingleNode || runtimeType === runtimeTypes.dataprocCluster;
+
+export const isDataprocCluster = (runtimeType) => runtimeType === runtimeTypes.dataprocCluster;
+
+// TODO: use type
+export const runtimeTypes = {
+  gceVm: 'Standard VM',
+  dataprocSingleNode: 'Spark single node',
+  dataprocCluster: 'Spark cluster',
+};
+
+export const shouldUsePersistentDisk = (runtimeType, runtimeDetails, upgradeDiskSelected) =>
+  isGce(runtimeType) && (!runtimeDetails?.runtimeConfig?.diskSize || upgradeDiskSelected);
+
+// TODO: tnis is the best place to start
 export const buildExistingEnvironmentConfig = (
   computeConfig: IComputeConfig,
-  currentRuntimeDetails: any,
-  currentPersistentDiskDetails: PersistentDisk,
-  isDataproc: boolean
-) => {
+  currentRuntimeDetails?: GetRuntimeItem,
+  currentPersistentDiskDetails?: PersistentDisk,
+  isDataproc = false
+): ExistingModalRuntimeConfig => {
   const runtimeConfig = currentRuntimeDetails?.runtimeConfig;
   const cloudService = runtimeConfig?.cloudService;
-  const numberOfWorkers = runtimeConfig?.numberOfWorkers || 0;
-  const hasGpu = computeConfig.hasGpu;
-  const gpuConfig = runtimeConfig?.gpuConfig;
-  const toolLabel = getToolLabelFromCloudEnv(currentRuntimeDetails);
+  const numberOfWorkers = runtimeConfig && 'numberOfWorkers' in runtimeConfig ? runtimeConfig.numberOfWorkers : 0;
+  const gpuConfig = runtimeConfig && 'gpuConfig' in runtimeConfig ? runtimeConfig.gpuConfig : undefined;
+  const toolLabel = currentRuntimeDetails ? getToolLabelFromCloudEnv(currentRuntimeDetails) : undefined;
   return {
-    hasGpu,
+    hasGpu: computeConfig.hasGpu,
     autopauseThreshold: currentRuntimeDetails?.autopauseThreshold,
+    computeConfig,
+    currentRuntimeDetails,
     runtime: currentRuntimeDetails
       ? {
           cloudService,
           toolDockerImage: getImageUrlFromRuntime(currentRuntimeDetails),
           tool: toolLabel,
+          region: computeConfig.computeRegion,
+          persistentDiskAttached: !!(runtimeConfig && 'persistentDiskId' in runtimeConfig),
+          gpuConfig,
           ...(currentRuntimeDetails?.jupyterUserScriptUri
             ? {
                 jupyterUserScriptUri: currentRuntimeDetails?.jupyterUserScriptUri,
               }
             : {}),
-          ...(currentRuntimeDetails?.timeoutInMinutes
-            ? {
-                timeoutInMinutes: currentRuntimeDetails?.timeoutInMinutes,
-              }
-            : {}),
+          // TODO: this can use runtimeconfig conditionals too more safely
           ...(cloudService === cloudServices.GCE
             ? {
                 zone: computeConfig.computeZone,
-                region: computeConfig.computeRegion,
```

Comment:
So no need of the region anymore? Only the zone was used?

Reply:
the region was duplicated on line 69 and 82 in both branches of the conditional, I simplified by moving it up to here https://github.com/DataBiosphere/terra-ui/pull/5079/files#diff-32a05189ab14623d039151b28dfee24908fc62cad51bf92afdb7767afbf7ef01R128